Parsing totally breaks when functions containing type arguments are defined "above" a class, but not below.

Originally reported here: https://github.com/typed-ember/glint/issues/408

```
❯ cat package.json | grep \"ember-template-imports
    "ember-template-imports": "^3.1.1",
```

<details><summary>No type errors</summary>

```ts
import Component from '@glimmer/component';
import { cached, tracked } from '@glimmer/tracking';
import { assert } from '@ember/debug';

import * as d3 from 'd3';

import { Info, type SunburstData } from './info';
import { autosize } from './autosize';

interface Signature {
  Args: {
    data: Info;
  }
}

interface Data {
  name: string;
  value?: number;
}

export class Sunburst extends Component<Signature> {
  <template>
    <svg {{autosize this.size this.updateSize}} style="width: 100%; height: 100%;">
      <g fill-opacity={{0.6}}>
        {{#each (descendentsWithDepth this.root) as |node|}}
          <path fill={{this.getColor node}} d={{this.arc node}}>
            <text>{{nodeText node}}</text>
          </path>
        {{/each}}
      </g>
      <g pointer-events="none" text-anchor="middle" font-size=10 font-family="sans-serif">
        {{#each (descendentsWithReportableSize this.root) as |node|}}
          <text transform={{this.getTextTransform node}} dy="0.35em">{{node.data.name}}</text>
        {{/each}}
      </g>
    </svg>
  </template>

  @tracked size = 700;

  updateSize = (size: number) => this.size = size;

  get radius() {
    return this.size / 2;
  }

  get data() {
    return this.args.data.json;
  }

  @cached
  get root() {
    return partition(this.data, this.radius);
  }

  /**
    * Cached function - provides a color-getting function based on the number of
    * root childern
    */
  @cached
  get color() {
    return d3.scaleOrdinal(
      d3.quantize(d3.interpolateRainbow, this.data.children.length + 1)
    );
  }


  /**
    * Cached function - provides an arc-getting function based on the radius
    */
  @cached
  get arc() {
    return d3
      .arc<d3.HierarchyRectangularNode<Data>>()
      .startAngle((d) => d.x0)
      .endAngle((d) => d.x1)
      .padAngle((d) => Math.min((d.x1 - d.x0) / 2, 0.005))
      .padRadius(this.radius / 2)
      .innerRadius((d) => d.y0)
      .outerRadius((d) => d.y1 - 1);
  }

  getColor = (d: d3.HierarchyRectangularNode<Data>) => {
    let currentNode: null | d3.HierarchyRectangularNode<Data> = d;
    while (currentNode && currentNode.depth > 1) currentNode = currentNode.parent;

    assert(`Search for the parent node failed`, currentNode);

    return this.color(currentNode.data.name);
  };

  getTextTransform = (d: d3.HierarchyRectangularNode<Data>) => {
    const x = (((d.x0 + d.x1) / 2) * 180) / Math.PI;
    const y = (d.y0 + d.y1) / 2;

    return `rotate(${x - 90}) translate(${y},0) rotate(${x < 180 ? 0 : 180})`;
  };
}

const format = d3.format(",d");

function descendentsWithDepth(root: d3.HierarchyRectangularNode<Data>) {
  return root.descendants().filter(d => d.depth);
}

function descendentsWithReportableSize(root: d3.HierarchyRectangularNode<Data>) {
  return root
    .descendants()
    .filter((d) => d.depth && ((d.y0 + d.y1) / 2) * (d.x1 - d.x0) > 10);
}

function nodeText(node: d3.HierarchyRectangularNode<Data>) {
  return node
    .ancestors()
    .map((d) => d.data.name)
    .reverse()
    .join("/") + '\n' + format(node.value ?? 0);
}

function partition(data: SunburstData, radius: number) {
  return d3.partition<Data>().size([2 * Math.PI, radius])(
    d3
      .hierarchy(data)
      .sum((d) => d.value ?? 0)
      .sort((a, b) => (b.value ?? 0) - (a.value ?? 0))
  );
}

```

</details>
<details><summary>Type errors</summary>

```ts
import Component from '@glimmer/component';
import { cached, tracked } from '@glimmer/tracking';
import { assert } from '@ember/debug';

import * as d3 from 'd3';

import { Info, type SunburstData } from './info';
import { autosize } from './autosize';

interface Signature {
  Args: {
    data: Info;
  }
}

interface Data {
  name: string;
  value?: number;
}


function descendentsWithReportableSize(root: d3.HierarchyRectangularNode<Data>) {
  return root
    .descendants()
    .filter((d) => d.depth && ((d.y0 + d.y1) / 2) * (d.x1 - d.x0) > 10);
}

export class Sunburst extends Component<Signature> {
  <template>
    <svg {{autosize this.size this.updateSize}} style="width: 100%; height: 100%;">
      <g fill-opacity={{0.6}}>
        {{#each (descendentsWithDepth this.root) as |node|}}
          <path fill={{this.getColor node}} d={{this.arc node}}>
            <text>{{nodeText node}}</text>
          </path>
        {{/each}}
      </g>
      <g pointer-events="none" text-anchor="middle" font-size=10 font-family="sans-serif">
        {{#each (descendentsWithReportableSize this.root) as |node|}}
          <text transform={{this.getTextTransform node}} dy="0.35em">{{node.data.name}}</text>
        {{/each}}
      </g>
    </svg>
  </template>

  @tracked size = 700;

  updateSize = (size: number) => this.size = size;

  get radius() {
    return this.size / 2;
  }

  get data() {
    return this.args.data.json;
  }

  @cached
  get root() {
    return partition(this.data, this.radius);
  }

  /**
    * Cached function - provides a color-getting function based on the number of
    * root childern
    */
  @cached
  get color() {
    return d3.scaleOrdinal(
      d3.quantize(d3.interpolateRainbow, this.data.children.length + 1)
    );
  }


  /**
    * Cached function - provides an arc-getting function based on the radius
    */
  @cached
  get arc() {
    return d3
      .arc<d3.HierarchyRectangularNode<Data>>()
      .startAngle((d) => d.x0)
      .endAngle((d) => d.x1)
      .padAngle((d) => Math.min((d.x1 - d.x0) / 2, 0.005))
      .padRadius(this.radius / 2)
      .innerRadius((d) => d.y0)
      .outerRadius((d) => d.y1 - 1);
  }

  getColor = (d: d3.HierarchyRectangularNode<Data>) => {
    let currentNode: null | d3.HierarchyRectangularNode<Data> = d;
    while (currentNode && currentNode.depth > 1) currentNode = currentNode.parent;

    assert(`Search for the parent node failed`, currentNode);

    return this.color(currentNode.data.name);
  };

  getTextTransform = (d: d3.HierarchyRectangularNode<Data>) => {
    const x = (((d.x0 + d.x1) / 2) * 180) / Math.PI;
    const y = (d.y0 + d.y1) / 2;

    return `rotate(${x - 90}) translate(${y},0) rotate(${x < 180 ? 0 : 180})`;
  };
}

const format = d3.format(",d");

function descendentsWithDepth(root: d3.HierarchyRectangularNode<Data>) {
  return root.descendants().filter(d => d.depth);
}


function nodeText(node: d3.HierarchyRectangularNode<Data>) {
  return node
    .ancestors()
    .map((d) => d.data.name)
    .reverse()
    .join("/") + '\n' + format(node.value ?? 0);
}

function partition(data: SunburstData, radius: number) {
  return d3.partition<Data>().size([2 * Math.PI, radius])(
    d3
      .hierarchy(data)
      .sum((d) => d.value ?? 0)
      .sort((a, b) => (b.value ?? 0) - (a.value ?? 0))
  );
}

```

</details>

The only difference is that _to get things working_, I move the functions _underneath_ the class.



_specifically_, this function:
```ts
function descendentsWithReportableSize(root: d3.HierarchyRectangularNode<Data>) {
  return root
    .descendants()
    .filter((d) => d.depth && ((d.y0 + d.y1) / 2) * (d.x1 - d.x0) > 10);
}
```
when it exists above the class causes _everything_ to freak out :scream: lol

Here is what it looks like in the editor:
![image](https://user-images.githubusercontent.com/199018/189545459-5bce3c40-638e-4d9b-92b2-023e34008e4c.png)

<details><summary>about 1/4th of the output of running the glint CLI (there is a lot)</summary>

It seems every identifier has an error?

```
❯ pnpm glint
app/components/ram-analyzer/sunburst.gts:63:4 - error TS2304: Cannot find name 'template'.

63   <template>
      ~~~~~~~~

app/components/ram-analyzer/sunburst.gts:64:6 - error TS2304: Cannot find name 'svg'.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
        ~~~

app/components/ram-analyzer/sunburst.gts:64:10 - error TS1005: '>' expected.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
            ~

app/components/ram-analyzer/sunburst.gts:64:11 - error TS1136: Property assignment expected.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
             ~

app/components/ram-analyzer/sunburst.gts:64:12 - error TS1434: Unexpected keyword or identifier.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
              ~~~~~~~~

app/components/ram-analyzer/sunburst.gts:64:21 - error TS2304: Cannot find name 'initial'.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
                       ~~~~~~~

app/components/ram-analyzer/sunburst.gts:64:29 - error TS2532: Object is possibly 'undefined'.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
                               ~~~~

app/components/ram-analyzer/sunburst.gts:64:39 - error TS1128: Declaration or statement expected.

64     <svg {{autosize initial=this.size}} style="width: 100%; height: 100%;">
```

</details>

--- a/__tests__/parse-templates.test.ts
+++ b/__tests__/parse-templates.test.ts
@@ -62,6 +62,91 @@ describe('parseTemplates', function () {
     `);
   });
 
+  it('<template></template> preceded by a slash character', function () {
+    const input = `
+      const divide = () => 4 / 2;
+      <template>Hello!</template>
+    `;
+
+    const templates = parseTemplates(input, 'foo.gjs', {
+      templateTag: 'template',
+    });
+
+    expect(templates).toMatchInlineSnapshot(`
+      Array [
+        Object {
+          "end": Object {
+            "0": "</template>",
+            "1": undefined,
+            "groups": undefined,
+            "index": 57,
+            "input": "
+            const divide = () => 4 / 2;
+            <template>Hello!</template>
+          ",
+          },
+          "start": Object {
+            "0": "<template>",
+            "1": undefined,
+            "groups": undefined,
+            "index": 41,
+            "input": "
+            const divide = () => 4 / 2;
+            <template>Hello!</template>
+          ",
+          },
+          "type": "template-tag",
+        },
+      ]
+    `);
+  });
+
+  // This test demonstrates a problem with the current implementation. The characters "<template>"
+  // inside of a regex is treated as an opening template tag instead of being ignored. Previous
+  // versions of this addon attempted to address this by parsing "/"-delimited regexes, however
+  // the addon's regular-expression based tokenizing was unable to properly distinguish a regex
+  // from division, and so the test above this one ("<template></template> preceded by a slash
+  // character") did not pass and caused quite confusing failures that occurred far more
+  // frequently than the issue demonstrated below will occur.
+  it.skip('<template></template> with <template> inside of a regexp', function () {
+    const input = `
+      const myregex = /<template>/;
+      <template>Hello!</template>
+    `;
+
+    const templates = parseTemplates(input, 'foo.gjs', {
+      templateTag: 'template',
+    });
+
+    expect(templates).toMatchInlineSnapshot(`
+      Array [
+        Object {
+          "end": Object {
+            "0": "</template>",
+            "1": undefined,
+            "groups": undefined,
+            "index": 9,
+            "input": "
+            const myregex = /<template>/;
+            <template>Hello!</template>
+          ",
+          },
+          "start": Object {
+            "0": "<template>",
+            "1": undefined,
+            "groups": undefined,
+            "index": 43,
+            "input": "
+            const myregex = /<template>/;
+            <template>Hello!</template>
+          ",
+          },
+          "type": "template-tag",
+        },
+      ]
+    `);
+  });
+
   it('hbs`Hello!` when only matching <template>', function () {
     const input = 'hbs`Hello!`';
 

--- a/src/parse-templates.ts
+++ b/src/parse-templates.ts
@@ -49,7 +49,7 @@ export interface ParseTemplatesOptions {
 }
 
 const escapeChar = '\\';
-const stringOrRegexDelimiter = /['"/]/;
+const stringDelimiter = /['"]/;
 
 const singleLineCommentStart = /\/\//;
 const newLine = /\n/;
@@ -119,7 +119,7 @@ export function parseTemplates(
       newLine.source,
       multiLineCommentStart.source,
       multiLineCommentEnd.source,
-      stringOrRegexDelimiter.source,
+      stringDelimiter.source,
       templateLiteralStart.source,
       templateLiteralEnd.source,
       dynamicSegmentStart.source,
@@ -170,16 +170,16 @@ export function parseTemplates(
       token[0].match(templateTagStart)
     ) {
       parseTemplateTag(results, template, token, tokens);
-    } else if (token[0].match(stringOrRegexDelimiter)) {
-      parseStringOrRegex(results, template, token, tokens);
+    } else if (token[0].match(stringDelimiter)) {
+      parseString(results, template, token, tokens);
     }
   }
 
   /**
-   * Parse a string or a regex. All tokens within a string or regex are ignored
+   * Parse a string. All tokens within a string are ignored
    * since there are no dynamic segments within these.
    */
-  function parseStringOrRegex(
+  function parseString(
     _results: TemplateMatch[],
     template: string,
     startToken: RegExpMatchArray,

--- a/tests/integration/gjs-test.gjs
+++ b/tests/integration/gjs-test.gjs
@@ -52,8 +52,26 @@ module('tests/integration/components/gjs', function (hooks) {
         scope: () => ({ Foo }),
       })
     );
-
     assert.equal(this.element.textContent.trim(), 'Hello, `lifeform`!');
+  });
+
+test('it works with classes with a slash character somewhere before the template', async function (assert) {
+    class Foo extends Component {
+      greeting = 'Hello';
+      get age() {
+        return 90 / 2;
+      }
+
+      <template>{{this.greeting}}, {{this.age}}-year-old!</template>
+    }
+
+    await render(
+      precompileTemplate(`<Foo @name="world" />`, {
+        strictMode: true,
+        scope: () => ({ Foo }),
+      })
+    );
+    assert.equal(this.element.textContent.trim(), 'Hello, 45-year-old!');
   });
 
   test('it works with a component that is a top-level default export', async function (assert) {

